lib/google/gax/operation.rb

```diff
@@ -0,0 +1,163 @@
+require 'google/gax/constants'
+require 'google/gax/settings'
+require 'google/longrunning/operations_api'
+
+module Google
+  module Gax
+    # A class used to wrap Google::Longrunning::Operation objects.
+    #
+    # @example General usage example
+    #   require 'google/gax/operation'
+    #
+    #   op = Google::Gax::Operation.new api.methodThatReturnsOperation(name)
+    #   op.done?
+    #   op.reload!
+    #   op.done?
+    #   results = op.results
+    #   raise results if op.error?
+    #
+    # @attribute [r] grpc_op
+    #   @return [Google::Longrunning::Operation] The wrapped grpc
+    #     operation object.
+    # @attribute [r] client
+    #   @return [Google::Longrunning::OperationsApi] The client that handles the
+    #     grpc operations.
+    class Operation
+      attr_reader :grpc_op, :client
+
+      # @param grpc_op [Google::Longrunning::Operation]
+      #   The inital longrunning operation.
+      # @param client [Google::Longrunning::OperationsApi]
+      #   The client that handles the grpc operations.
+      def initialize(grpc_op, client: nil)
+        @grpc_op = grpc_op
+        @client = client ? client : Google::Longrunning::OperationsApi.new
+        @callbacks = []
+      end
+
+      # If the operation is done, returns the result, otherwise returns nil.
+      # If a Class is provided, an instance of that class will try to be
+      # unpacked from the response. If the response is not of the type provided
+      # nil will be returned.
+      #
+      # @param [Class] The class type to be unpacked from the response.
+      # @return [nil | Google::Rpc::Status | Object | Google::Protobuf::Any ]
+      #   The result of the operation
+      def results(responseType: nil)
+        return nil unless done?
+        return @grpc_op.error if error?
+        return @grpc_op.response.unpack(responseType) if responseType
+        @grpc_op.response
+      end
+
+      # Checks if the operation is done. This does not send a new api call,
+      # but checks the result of the previous api call to see if done.
+      #
+      # @return [Boolean] Whether the operation is done.
+      def done?
+        @grpc_op.done
+      end
+
+      # Checks if the operation is done and the result is an error.
+      # If the operation is not finished then this will return false.
+      #
+      # @return [Boolean] Whether an error has been returned.
+      def error?
+        done? ? @grpc_op.result == :error : false
+      end
+
+      # Reloads the operation object.
+      # @return [Google::Gax::Longrunning]
+      #   Since this method changes internal state, it returns itself.
+      def reload!(backoff_settings: nil)
+        options = CallSettings.new(
+          retry_options: RetryOptions.new(
+            backoff_settings: backoff_settings
+          )
+        )
+        @grpc_op = @client.get_operation @grpc_op.name, options: options
+        if done?
+          callbacks.each { |proc| proc.call(results) }
```

Comment:
Yep, that's my concern. There is also meta data that we can't infer from the results, such as operation id/name, that might be useful as well.